[2/3] Remove computed GOTOs

jkeiser opened the pull request:
Second in a three-part PR series: #1065 (anonymous namespaces) -- 1078 (remove computed GOTOs) -- #1079 (return errors instead of goto).

This:
* Replaces all computed gotos with static gotos, changing actual flow as minimally as possible.
* Removes unified_machine_addresses, ret_address_t, and parse_value(), simplifying the parse_value switch to keep the main loop as simple as possible.
* Replaces the stack of continue addresses with a stack of booleans indicating whether each open scope is an object or an array.
* Marks parse_structurals() (the method with the gotos) as really_inline, indicating the inlining infrastructure may be able to work with our code better now that we're not using dynamic gotos.

The primary goal is to simplify control flow and use fewer fancy features where we can. This is also a necessary step towards a clean SAX interface, separating stage 2 iteration from stage 2 write. Otherwise, callbacks will have to receive (and return) computed goto addresses to keep the same level of performance as simdjson's current stage 2.

## Performance

This is a minor loss on Intel. It fluctuates each time I make a change. Instruction count is almost randomly all over the place. I think it's jitters. I have no idea how to prove otherwise.

ARM gcc10 impact is a very minor loss for some files, but doesn't make anything worse. Instruction count is all over the place, just like with Intel.

This does not appear to make Windows much faster (less than 1% for twitter.json on my machine, which is admittedly a laptop).

### skylake-x gcc9

| File                             | Blocks | master Cycles | PR Cycles | +Throughput | master Instr. | PR Instr. | -Instr. |
| ---                              |    --: |   --: |   --: |   --: |   --: |   --: |   --: |
| mesh.pretty                      |  24646 | 184.7 | 180.3 |    2% | 584.1 | 579.2 |    0% |
| canada                           |  35172 | 295.9 | 289.1 |    2% | 982.0 | 982.2 |    0% |
| twitterescaped                   |   8787 | 190.8 | 190.5 |    0% | 512.0 | 515.5 |    0% |
| instruments                      |   3442 | 108.5 | 108.8 |    0% | 376.1 | 373.6 |    0% |
| twitter                          |   9867 |  91.6 |  92.6 |   -1% | 296.9 | 296.8 |    0% |
| citm_catalog                     |  26987 |  83.6 |  84.8 |   -1% | 312.5 | 314.5 |    0% |
| github_events                    |   1017 |  77.3 |  78.5 |   -1% | 267.3 | 269.4 |    0% |
| gsoc-2018                        |  51997 |  72.9 |  74.2 |   -1% | 169.0 | 169.8 |    0% |
| mesh                             |  11306 | 304.7 | 311.2 |   -2% | 989.7 | 978.9 |    1% |
| random                           |   7976 | 141.6 | 145.4 |   -2% | 496.8 | 497.3 |    0% |
| marine_ik                        |  46616 | 291.0 | 299.2 |   -2% | 915.6 | 912.6 |    0% |
| update-center                    |   8330 | 113.0 | 116.6 |   -3% | 342.6 | 344.8 |    0% |
| apache_builds                    |   1988 |  84.3 |  87.8 |   -4% | 307.4 | 311.4 |   -1% |
| numbers                          |   2345 | 241.1 | 253.4 |   -5% | 799.3 | 790.8 |    1% |

There is a great deal of apparent compiler-generated randomness here. For example, 

* [Make parse_structurals really_inline](https://github.com/simdjson/simdjson/pull/1078/commits/460cfcaf3e42fb86bc107dcc7c080dfeb71520ed) makes it jump around randomly but generally reduces instruction count:

  | File                             | Blocks | +Throughput | -Instr. |
  | ---                              |    --: |   --: |   --: |
  | marine_ik                        |  46616 |    4% |    0% |
  | numbers                          |   2345 |    4% |    0% |
  | citm_catalog                     |  26987 |    3% |    2% |
  | mesh                             |  11306 |    3% |    1% |
  | mesh.pretty                      |  24646 |    2% |    0% |
  | canada                           |  35172 |    0% |    0% |
  | twitterescaped                   |   8787 |    0% |    0% |
  | apache_builds                    |   1988 |    0% |    0% |
  | update-center                    |   8330 |    0% |    0% |
  | gsoc-2018                        |  51997 |   -1% |    0% |
  | github_events                    |   1017 |   -1% |    0% |
  | random                           |   7976 |   -1% |    0% |
  | twitter                          |   9867 |   -2% |    0% |
  | instruments                      |   3442 |   -5% |    0% |

### ARM gcc9

| File                             | Blocks | master Cycles | PR Cycles | +Throughput | master Instr. | PR Instr. | -Instr. | master Branch Misses | PR Branch Misses | -Branch Misses | master Cache Misses | PR Cache Misses | -Cache Misses | master Cache Refs | PR Cache Refs | -Cache Refs |
| ---                              |    --: |   --: |   --: |   --: |   --: |   --: |   --: |    --: |    --: |   --: |    --: |    --: |   --: |    --: |    --: |   --: |
| canada                           |  35172 | 852.6 | 837.5 |    1% | 1018.3 | 1026.4 |    0% |  52366 |  57395 |   -9% |  91681 |  91733 |    0% | 4904465 | 5103335 |   -4% |
| citm_catalog                     |  26987 | 397.7 | 393.1 |    1% | 370.6 | 373.4 |    0% |   4841 |   4702 |    2% |  62736 |  62727 |    0% | 1924043 | 1959557 |   -1% |
| mesh                             |  11306 | 949.4 | 940.4 |    0% | 1062.2 | 1054.9 |    0% |  21073 |  20218 |    4% |  32325 |  32344 |    0% | 1622781 | 1683179 |   -3% |
| marine_ik                        |  46616 | 893.6 | 886.1 |    0% | 988.2 | 987.6 |    0% |  99457 | 100457 |   -1% | 134021 | 133966 |    0% | 7439148 | 7781571 |   -4% |
| mesh.pretty                      |  24646 | 597.5 | 593.5 |    0% | 635.0 | 631.6 |    0% |  39558 |  40186 |   -1% |  59084 |  59080 |    0% | 1971802 | 2032827 |   -3% |
| instruments                      |   3442 | 482.0 | 479.5 |    0% | 462.8 | 462.6 |    0% |   3556 |   3215 |    9% |   8696 |   8674 |    0% | 335597 | 338701 |    0% |
| apache_builds                    |   1988 | 472.7 | 470.2 |    0% | 414.3 | 418.2 |    0% |    657 |    628 |    4% |   4836 |   4832 |    0% | 201636 | 203258 |    0% |
| gsoc-2018                        |  51997 | 360.1 | 360.3 |    0% | 238.8 | 239.4 |    0% |  38814 |  39638 |   -2% | 109478 | 109696 |    0% | 2755135 | 2763570 |    0% |
| twitter                          |   9867 | 489.4 | 489.7 |    0% | 386.9 | 387.2 |    0% |  16007 |  15477 |    3% |  23437 |  23435 |    0% | 871899 | 871078 |    0% |
| twitterescaped                   |   8787 | 736.7 | 737.3 |    0% | 629.4 | 629.8 |    0% |  19334 |  17907 |    7% |  21356 |  21356 |    0% | 1574717 | 1572639 |    0% |
| update-center                    |   8330 | 546.5 | 547.8 |    0% | 463.1 | 465.8 |    0% |  13723 |  13289 |    3% |  20804 |  20805 |    0% | 1037169 | 1033504 |    0% |
| random                           |   7976 | 701.8 | 703.7 |    0% | 639.4 | 643.1 |    0% |   8593 |   6693 |   22% |  21618 |  21609 |    0% | 1250791 | 1249231 |    0% |
| numbers                          |   2345 | 907.1 | 909.8 |    0% | 874.0 | 869.7 |    0% |   3283 |   3268 |    0% |   6030 |   6015 |    0% | 239948 | 259502 |   -8% |
| github_events                    |   1017 | 460.6 | 462.2 |    0% | 358.8 | 359.9 |    0% |   1523 |   1553 |   -1% |   2412 |   2406 |    0% |  88367 |  88624 |    0% |

lemire commented:
@jkeiser Odd that it would not help Visual Studio.

lemire commented:
Compared to master, my own tests do not indicate a performance regression.

lemire commented on src/arm64/bitmanipulation.h:
```diff
@@ -55,6 +55,7 @@ really_inline bool add_overflow(uint64_t value1, uint64_t value2, uint64_t *resu
 #endif
 }
 
+#if 0 // Currently unused
```
Maybe we could just delete this code section. We can always retrieve it from previous releases.

lemire commented on src/arm64/bitmanipulation.h:
```diff
@@ -63,8 +64,9 @@ really_inline bool mul_overflow(uint64_t value1, uint64_t value2, uint64_t *resu
 
 } // namespace arm64
-} // namespace simdjson
+} // namespace {
```
Though `namespace {` is fine, I think that the technical term might be "unnamed namespace". If you want to sound more scientific.

https://en.cppreference.com/w/cpp/language/namespace#Unnamed_namespaces

lemire commented on src/fallback/bitmanipulation.h:
```diff
@@ -24,9 +24,10 @@ static unsigned char _BitScanReverse64(unsigned long* ret, uint64_t x) {
+//
+// These are currently unused, but one day will be.
+//
+#if 0 // Currently unused
```
I'd probably just remove the code, we can get it back from earlier versions.

lemire commented on src/generic/stage2/allocate.h:
```diff
@@ -8,9 +8,9 @@ namespace allocate {
 really_inline error_code set_max_depth(dom_parser_implementation &parser, size_t max_depth) {
   parser.containing_scope.reset(new (std::nothrow) scope_descriptor[max_depth]);
-  parser.ret_address.reset(new (std::nothrow) ret_address_t[max_depth]);
+  parser.is_array.reset(new (std::nothrow) bool[max_depth]);
```
One nice benefit of this change is that `bool` typically uses one one byte unlike the addresses which would use 8 bytes. It would be possible to use just one bit but at the expense of some work (vector<bool> might implement this 1-bit trade-off).

jkeiser replied:
Yeah. In fact, 99.9% of json files could be handled with a single 64-bit integer. (Honestly 32 would be more than enough.) I wasn't quite ready to go that crazy, but it wouldn't be terribly difficult, I think.

lemire commented on src/haswell/bitmanipulation.h:
```diff
@@ -53,6 +53,7 @@ really_inline bool add_overflow(uint64_t value1, uint64_t value2,
 #endif
 }
 
+#if 0 // Currently unused
```
If we don't use code we should probably delete it.

lemire commented on src/westmere/bitmanipulation.h:
```diff
@@ -62,6 +62,7 @@ really_inline bool add_overflow(uint64_t value1, uint64_t value2,
 #endif
 }
 
+#if 0 // Currently unused
```
Same comment. If we don't use something, let us kill it.

lemire commented:
One reason why this PR works well performance-wise is that the new switch/case constructs are compiled systematically into tight lookup tables...

https://godbolt.org/z/53MTGY

The clang compiler seems to do the best job with it but all compilers appear to deal with this code nicely.

lemire commented:
I have converted the PR to a draft because I want @jkeiser  to consider adding an "unlikely" keyword on one conditional. It could change the performance outlook, at least for some cases.

jkeiser commented:
@lemire I added the two `unlikely` branches and removed the ` else { }` from `else { goto }`, and this is now a definite improvement against master (updated numbers at the top). Further, I applied the `unlikely` changes to master (which gives a small improvement there), and diffed against that. I found that this PR is *still* a benefit, even discounting the `unlikely` effect:

| File                             | Blocks | master Cycles | PR Cycles | +Throughput | master Instr. | PR Instr. | -Instr. | master Branch Misses | PR Branch Misses | -Branch Misses | master Cache Misses | PR Cache Misses | -Cache Misses | master Cache Refs | PR Cache Refs | -Cache Refs |
| ---                              |    --: |   --: |   --: |   --: |   --: |   --: |   --: |    --: |    --: |   --: |    --: |    --: |   --: |    --: |    --: |   --: |
| marine_ik                        |  46616 | 295.5 | 282.0 |    4% | 909.1 | 922.4 |   -1% |  53973 |  53071 |    1% | 111557 | 111216 |    0% | 255870 | 255770 |    0% |
| mesh                             |  11306 | 307.8 | 295.0 |    4% | 984.6 | 991.9 |    0% |  10430 |  10295 |    1% |      8 |     66 | -725% |  60003 |  60853 |   -1% |
| instruments                      |   3442 | 104.4 | 100.8 |    3% | 377.4 | 374.6 |    0% |    399 |    517 |  -29% |     15 |      1 |   93% |    910 |    940 |   -3% |
| mesh.pretty                      |  24646 | 180.4 | 177.3 |    1% | 584.2 | 586.8 |    0% |  13942 |  13811 |    0% |    663 |    466 |   29% |  88013 |  88189 |    0% |
| numbers                          |   2345 | 254.2 | 251.0 |    1% | 798.1 | 802.3 |    0% |   2122 |   2138 |    0% |      0 |      0 |    0% |    322 |      5 |   98% |
| gsoc-2018                        |  51997 |  74.7 |  73.8 |    1% | 171.4 | 170.9 |    0% |  30174 |  30490 |   -1% |  55826 |  53100 |    4% | 168496 | 168384 |    0% |
| citm_catalog                     |  26987 |  86.6 |  85.8 |    0% | 319.5 | 313.4 |    1% |   1816 |   1760 |    3% |   2184 |   1150 |   47% |  89540 |  89509 |    0% |
| update-center                    |   8330 | 114.7 | 114.5 |    0% | 343.3 | 342.2 |    0% |   6032 |   5688 |    5% |      2 |      4 | -100% |  33920 |  33080 |    2% |
| random                           |   7976 | 141.9 | 142.6 |    0% | 497.0 | 493.9 |    0% |   2099 |   2081 |    0% |    332 |    205 |   38% |  37557 |  36861 |    1% |
| github_events                    |   1017 |  77.9 |  78.3 |    0% | 270.1 | 267.8 |    0% |     84 |    105 |  -25% |      0 |      0 |    0% |      1 |      0 |    0% |
| twitterescaped                   |   8787 | 189.2 | 190.4 |    0% | 511.3 | 512.3 |    0% |   4742 |   5013 |   -5% |    159 |     24 |   84% |  30050 |  29048 |    3% |
| twitter                          |   9867 |  91.5 |  92.9 |   -1% | 299.8 | 297.4 |    0% |   3125 |   3465 |  -10% |    354 |    299 |   15% |  33785 |  31769 |    5% |
| canada                           |  35172 | 289.1 | 293.8 |   -1% | 976.7 | 994.6 |   -1% |  18685 |  18518 |    0% |  22833 |  21826 |    4% | 154751 | 154738 |    0% |
| apache_builds                    |   1988 |  84.5 |  86.0 |   -1% | 308.1 | 308.1 |    0% |    104 |    112 |   -7% |      0 |      0 |    0% |      0 |      0 |    0% |
